File: ios/Snag/Support/SnagBoundedQueue.swift
import Foundation

struct SnagQueueMetricsSnapshot {
    let queuedPackets: Int
    let droppedPackets: Int
    let enqueuedPackets: Int
}

final class SnagBoundedQueue<Element> {
    private let maxSize: Int
    private var items: [Element] = []
    private var droppedCount: Int = 0
    private var enqueuedCount: Int = 0
    private let lock = NSLock()

    init(maxSize: Int) {
        self.maxSize = max(1, maxSize)
    }

    @discardableResult
    func enqueue(_ item: Element) -> Bool {
        lock.lock(); defer { lock.unlock() }
        enqueuedCount += 1
        var dropped = false
        if items.count >= maxSize {
            items.removeFirst()
            droppedCount += 1
            dropped = true
        }
        items.append(item)
        return dropped
    }

    func drain() -> [Element] {
        lock.lock(); defer { lock.unlock() }
        let drained = items
        items.removeAll()
        return drained
    }

    func clear() {
        lock.lock(); defer { lock.unlock() }
        items.removeAll()
    }

    func snapshot() -> SnagQueueMetricsSnapshot {
        lock.lock(); defer { lock.unlock() }
        return SnagQueueMetricsSnapshot(
            queuedPackets: items.count,
            droppedPackets: droppedCount,
            enqueuedPackets: enqueuedCount
        )
    }
}

File: android/snag/src/main/java/com/snag/network/SnagBoundedQueue.kt
package com.snag.network

internal data class SnagBoundedQueueSnapshot(
    val size: Int,
    val droppedCount: Long,
    val enqueuedCount: Long
)

internal data class SnagBoundedQueueEnqueueResult(
    val dropped: Boolean,
    val size: Int,
    val droppedCount: Long
)

internal class SnagBoundedQueue<T>(private val maxSize: Int) {
    private val queue = ArrayDeque<T>()
    private var droppedCount: Long = 0
    private var enqueuedCount: Long = 0

    init {
        require(maxSize > 0) { "maxSize must be greater than zero" }
    }

    @Synchronized
    fun enqueue(item: T): SnagBoundedQueueEnqueueResult {
        enqueuedCount += 1
        var dropped = false
        if (queue.size >= maxSize) {
            queue.removeFirst()
            droppedCount += 1
            dropped = true
        }
        queue.addLast(item)
        return SnagBoundedQueueEnqueueResult(
            dropped = dropped,
            size = queue.size,
            droppedCount = droppedCount
        )
    }

    @Synchronized
    fun drain(): List<T> {
        val drained = queue.toList()
        queue.clear()
        return drained
    }

    @Synchronized
    fun size(): Int = queue.size

    @Synchronized
    fun snapshot(): SnagBoundedQueueSnapshot {
        return SnagBoundedQueueSnapshot(
            size = queue.size,
            droppedCount = droppedCount,
            enqueuedCount = enqueuedCount
        )
    }
}
